Jwt passport setup

lukasaric opened the pull request.

lukasaric committed:
Jwt passport setup

vladimyr commented on client/src/components/Header.vue:
```diff
@@ -88,7 +88,8 @@ export default {
       try {
-        this.$store.dispatch('clearUser', null);
+        await this.$store.dispatch('clearUser', null);
+        await JSON.parse(localStorage.removeItem('googleUser'));
```
Hmm, this doesn't make sense :confused: 
1. Typically, interacting with `localStorage` should be done inside `store`
2. `#removeItem` is synchronous so `await` here is useless
3. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Storage/removeItem#Return_value) return value of `#removeItem` is undefined so what exactly (and why) you are trying to deserialize aka parse from JSON? :confused:

vladimyr commented on client/src/store/store.js:
```diff
@@ -42,7 +42,7 @@ export default new Vuex.Store({
     },
     clearUser({commit}, user) {
-      commit('setUser', user);
+      commit('clearUser', user);
```
Isn't this exactly the same as commiting `setGoogleUser` with `null` as payload? :wink: 
Rethink this store in general...

vladimyr commented on server/controllers/AuthController.js:
```diff
@@ -3,6 +3,8 @@
 const GoogleStrategy = require('passport-google-oauth20');
 const LocalStrategy = require('passport-local').Strategy;
+const JwtStrategy = require('passport-jwt').Strategy;
+const ExtractJwt = require('passport-jwt').ExtractJwt;
```
But why :upside_down_face:
```js
const { ExtractJwt, JwtStrategy } = require('passport-jwt');
```

vladimyr commented on server/controllers/AuthController.js:
```diff
@@ -14,6 +16,17 @@ passport.deserializeUser((id, done) => {
 });
 
+passport.use(new JwtStrategy({
+  jwtFromRequest: ExtractJwt.fromHeader('authorization'),
```
Which scheme? Go read [source code](https://github.com/themikenicholson/passport-jwt/blob/bafdf99befe50fa6c90f982b4f92ce71a94b3ae1/lib/extract_jwt.js#L16-L24) and tell me does it use `JWT <token>` format, or `Bearer <token>` format or maybe just `<token>`? :thinking:

vladimyr replied:
Hint: There is better utility provided by `passport-jwt` and available of the shelf that you can use for this one...

lukasaric committed:
Login/Register JWT auth

- Successfully generated token
- JWT auth implemented on other routes
- Modified files ( created passport-setup and configured auth
 routes and controller)

vladimyr commented:
Last commit message, last list item:
`- Modified files ( created passport-setup and configured auth
 routes and controller)`
Kill this extra space after paren with fire 😿

vladimyr commented on server/app.js:
```diff
@@ -8,26 +8,19 @@ const createTest = require('./routes/createTest');
-app.use('/profile', profileStats);
-app.use('/topic', createTest);
+app.use('/profile', passport.authenticate('jwt', { session: false }), profileStats);
+app.use('/topic', passport.authenticate('jwt', { session: false }), createTest);
```
```js
const authenticate = passport.authenticate('jwt', { session: false });

app.use('/profile', authenticate, profileStats);
app.use('/topic', authenticate, createTest);
```
![](https://media.giphy.com/media/OFou8KcpXioUg/200.gif)

vladimyr commented on server/controllers/AuthController.js:
```diff
@@ -1,79 +1,54 @@
+  return jwt.sign({
+    sub: user.id
+  }, process.env.JWT_SECRET, { expiresIn: '2d' });
+};
```
How about:
```js
function createToken(user, expiresIn = '2d') {
  return jwt.sign({ sub: user.id }, process.env.JWT_SECRET, { expiresIn });
}
```
:bowtie:
 
👍 for using `expiresIn` with `zeit/ms` string

vladimyr commented on server/passport-setup.js:
```diff
@@ -0,0 +1,78 @@
+}, (payload, done) => {
+  return User.findOne({ id: payload.sub })
+    .then(user => {
+      return done(null, user);
```
```js
.then(user => done(null, user))
```

vladimyr commented on server/passport-setup.js:
```diff
@@ -0,0 +1,78 @@
+        .then(user => done(null, user))
+        .catch(err => {
+          if (err.name === 'SequelizeValidationError') {
+            return done(null, false, { message: 'Invalid input form.' });
```
<img src="https://s3.amazonaws.com/com.twilio.prod.twilio-docs/images/Screen-Shot-2017-03-06-at-5.11.06-PM.width-808.png" width="480">

Instead, return this `User.create(...)` call and attach `.then` & `.catch` blocks to existing promise chain. :bowtie:

vladimyr commented on server/passport-setup.js:
```diff
@@ -0,0 +1,78 @@
+      if (!user) {
+        return done(null, false, { message: 'Incorrect username.' });
+      }
+      const isValidPassword = await user.comparePassword(password, user.password);
```
Well, passing `user.password` to `User`'s password doesn't really have sense, make it accept only received password.

lukasaric committed:
JWT auth

- code grammar modification

lukasaric committed:
Modified promise chain in register

lukasaric committed:
Register error messages fixed

lukasaric committed:
Added global error handler

lukasaric committed:
Fixed register response

lukasaric committed:
Google auth token created

lukasaric committed:
Configured store

- manipulating with tokens in store

lukasaric committed:
Modified store

- configured tokens on client
